### DISCUSSION TITLE
What's the best way to use a semantic function within a native function?

### DISCUSSION THREAD

**Discussion by Swimburger** [microsoft/semantic-kernel#2754] [Q&A] [answered]
I have a native function that needs to call a sematic function.
I got this to work by passing the kernel into the ctor of the native function class, then add an inline function which I'm storing in a field.
From the native function I call the inline function. This works, but since I have to register the inline function, I have to provide the plugin name and function name, which I'd like the consumer of the plugin do. 

```csharp
public class MyPlugin
{
    private readonly IKernel _kernel;
    private readonly ISKFunction _getCommonFolderPathFunction;

    public MyPlugin(IKernel kernel)
    {
        _kernel = kernel;
        _getCommonFolderPathFunction = GetCommonFolderPathFunction();
    }

    private ISKFunction GetCommonFolderPathFunction()
    {
        var prompt = "What is the path for the common folder  {{$commonFolderName}} " +
                     $"on operating platform {Environment.OSVersion.Platform.ToString()} " +
                     $"with user profile path {Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)}?";
        var promptConfig = new PromptTemplateConfig
        {
            Schema = 1,
            Type = "completion",
            Description = "Gets the path for the given common folder.",
            Completion =
            {
                MaxTokens = 500,
                Temperature = 0.0,
                TopP = 0.0,
                PresencePenalty = 0.0,
                FrequencyPenalty = 0.0
            },
            Input =
            {
                Parameters = new List<PromptTemplateConfig.InputParameter>
                {
                    new()
                    {
                        Name = "commonFolderName",
                        Description = "Name of the common folder"
                    }
                }
            }
        };

        var promptTemplate = new PromptTemplate(prompt, promptConfig, _kernel);
        var functionConfig = new SemanticFunctionConfig(promptConfig, promptTemplate);
        return _kernel.RegisterSemanticFunction("MyPluginInternal", "GetCommonFolderPath", functionConfig);
    }

    private async Task<string?> GetCommonFolderPath(string commonFolderName)
    {
        var variables = new ContextVariables
        {
            ["commonFolderName"] = commonFolderName
        };
        var context = await _getCommonFolderPathFunction.InvokeAsync(variables);
        var match = Regex.Match(context.Result, @"([a-zA-Z]?\:?[\/\\].*[\/\\][\S-[""'\.]]*)", RegexOptions.IgnoreCase);
        return match.Success ? match.Captures[0].Value : null;
    }

    [SKFunction, Description("Foobar")]
    public async Task<string> LocateFile(SKContext context)
    {
        var foo = await GetCommonFolderPath(commonFolderName);
        ...
    }
}
```

Is there a way to have the inline function be more in line with the native function?
Provide attributes or have a function that returns the config, and it gets registered along with the other functions in the plugin?

**Comment by noeltupas** [comment]
The only way I've gotten this to work is to manually pass in the running kernel instance:

```
// note they removed IKernel
public MyPlugin(Kernel kernel)
{
    _kernel = kernel;
    _getCommonFolderPathFunction = GetCommonFolderPathFunction();
}
```

But this also means I cannot register this Native class at the same time I'm building out my Kernel instance. Which means whatever your firing mechanism is, it will need to register your Native class at runtime:

` var plugin = kernel.CreatePluginFromObject(new MyPlugin(kernel), "MyPlugin");`

This is what we currently do for Plugins that serve and orchestrations of other Semantic Functions, where an individual code path (for us, a REST endpoint) triggers the above per request.

Note Singleton Kernels are not recommended (per documentation). Transient ones are wasteful, and I haven't been able to get Scoped ones working (though I may need to wrap it in an abstraction first, I don't know...I'll have to try it).

My preference would be that every Kernel built through the builder registers an instance to itself in its own service collection. Then the above would work more intuitively. However, note that what we're asking for is the Kernel to inject itself into a Native Function that itself is trying to call. We don't ask a Controller to inject an instance of itself to execute a Controller method...that's some weird circular dependency blackhole. 

Perhaps it is cleaner/better to pass the dependency in to the function call while keeping the Native class's state free of a Kernel instance?

**Reply by stephentoub** [reply, marked as answer]
Why do you need to pass the Kernel to the plugin's constructor?  If you have a method like:
```C#
[KernelFunction]
public void MyAmazingMethod(Kernel kernel) { ... }
```
the Kernel being used when that method is invoked will be automatically passed in as that Kernel argument.

**Reply by noeltupas** [reply]
OMG I didn't know this! Thank you!

**Reply by noeltupas** [reply]
Just tried. Worked well. Thanks again!

**Reply by stephentoub** [reply]
Glad you got it working.

**Reply by stephentoub** [reply]
Seems like better docs are needed here. There are various things you can have as arguments that the system will automatically fill in, with Kernel being one but not the only one. For reference, and so you can see what else is handled, here's where it happens:
https://github.com/microsoft/semantic-kernel/blob/88bdb1155e148c98a72b171400cdc0ad0687ddae/dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethod.cs#L305-L372

**Reply by noeltupas** [reply]
Wow the cancellation token too! That's awesome. Was wondering about that as well...we were persisting in state, I was going to pass through KernelArguments. Now that I see this...that's just pure GOLD~ 🪙🥇

**Reply by MithrilMan** [reply]
@stephentoub I'm wondering what's the purpose to have KernelArguments there as well, I thought it was to manually add "arguments" in a multistep scenario where autofunction was invoked, but I see that each function call has its own KernelArguments that forget any previous method, I think would be nice to have a proper way to carry contextual informations in a cooked way, what do you think?

**Comment by noeltupas** [comment]
> Seems like better docs are needed here. 

Yes. Going through SK documentation it's never explicitly state whether or not a Native Function is intended to run Kernel functions internally (clearly it can but DI was awkward). Since the abstraction (IKernel) was removed, I figured you guys decided against the concept of Native functions serving as workflow orchestrations (of other Semantic and/or Native functions). 

The above knowledge you've provided I think would be super valuable to readers who are looking at docs and may not necessarily be looking through SK code 😜👍
